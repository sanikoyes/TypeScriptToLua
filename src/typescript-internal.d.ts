import { DiagnosticsProducingTypeChecker } from "./transformation/context";

export {};

declare module "typescript" {
    function createDiagnosticReporter(system: System, pretty?: boolean): DiagnosticReporter;
    function createWatchStatusReporter(system: System, pretty?: boolean): WatchStatusReporter;

    interface System {
        setBlocking?(): void;
    }

    interface Statement {
        jsDoc?: JSDoc[];
    }

    interface Program {
        getCommonSourceDirectory(): string;
        getDiagnosticsProducingTypeChecker(): DiagnosticsProducingTypeChecker;
    }

    interface CompilerOptions {
        configFile?: TsConfigSourceFile;
        configFilePath?: string;
    }

    interface TypeChecker {
        getElementTypeOfArrayType(type: Type): Type | undefined;
    }
}
