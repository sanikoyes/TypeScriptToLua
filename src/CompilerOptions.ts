import * as ts from "typescript";
import * as diagnosticFactories from "./transpilation/diagnostics";

type OmitIndexSignature<T> = {
    [K in keyof T as string extends K ? never : number extends K ? never : K]: T[K];
};

export interface TransformerImport {
    transform: string;
    import?: string;
    after?: boolean;
    afterDeclarations?: boolean;
    type?: "program" | "config" | "checker" | "raw" | "compilerOptions";
    [option: string]: any;
}

export interface LuaPluginImport {
    name: string;
    import?: string;
    [option: string]: any;
}

export type CompilerOptions = OmitIndexSignature<ts.CompilerOptions> & {
    buildMode?: BuildMode;
    noImplicitSelf?: boolean;
    noHeader?: boolean;
    luaBundle?: string;
    luaBundleEntry?: string;
    luaTarget?: LuaTarget;
    luaLibImport?: LuaLibImportKind;
    sourceMapTraceback?: boolean;
    luaPlugins?: LuaPluginImport[];
    plugins?: Array<ts.PluginImport | TransformerImport>;
    tstlVerbose?: boolean;
    [option: string]: any;
};

export enum LuaLibImportKind {
    None = "none",
    Always = "always",
    Inline = "inline",
    Require = "require",
}

export enum LuaTarget {
    Universal = "universal",
    Lua51 = "5.1",
    Lua52 = "5.2",
    Lua53 = "5.3",
    Lua54 = "5.4",
    LuaJIT = "JIT",
}

export enum BuildMode {
    Default = "default",
    Library = "library",
}

export const isBundleEnabled = (options: CompilerOptions) =>
    options.luaBundle !== undefined && options.luaBundleEntry !== undefined;

export function validateOptions(options: CompilerOptions): ts.Diagnostic[] {
    const diagnostics: ts.Diagnostic[] = [];

    if (options.luaBundle && !options.luaBundleEntry) {
        diagnostics.push(diagnosticFactories.luaBundleEntryIsRequired());
    }

    if (options.luaBundle && options.luaLibImport === LuaLibImportKind.Inline) {
        diagnostics.push(diagnosticFactories.usingLuaBundleWithInlineMightGenerateDuplicateCode());
    }

    if (options.luaBundle && options.buildMode === BuildMode.Library) {
        diagnostics.push(diagnosticFactories.cannotBundleLibrary());
    }

    return diagnostics;
}
