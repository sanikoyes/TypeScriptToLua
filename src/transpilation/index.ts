import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";
import { parseConfigFileWithSystem } from "../cli/tsconfig";
import { CompilerOptions } from "../CompilerOptions";
import { createEmitOutputCollector, TranspiledFile } from "./output-collector";
import { EmitResult, Transpiler } from "./transpiler";

export { Plugin } from "./plugins";
export * from "./transpile";
export * from "./transpiler";
export { EmitHost } from "./utils";
export { TranspiledFile };

export function transpileFiles(
    rootNames: string[],
    options: CompilerOptions = {},
    writeFile?: ts.WriteFileCallback
): EmitResult {
    const program = ts.createProgram(rootNames, options);
    const { diagnostics: transpileDiagnostics, emitSkipped } = new Transpiler().emit({ program, writeFile });
    const diagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...transpileDiagnostics,
    ]);

    return { diagnostics: [...diagnostics], emitSkipped };
}

export function transpileProject(
    configFileName: string,
    optionsToExtend?: CompilerOptions,
    writeFile?: ts.WriteFileCallback
): EmitResult {
    const parseResult = parseConfigFileWithSystem(configFileName, optionsToExtend);
    if (parseResult.errors.length > 0) {
        return { diagnostics: parseResult.errors, emitSkipped: true };
    }

    return transpileFiles(parseResult.fileNames, parseResult.options, writeFile);
}

const libCache: { [key: string]: ts.SourceFile } = {};

/** @internal */
export function createVirtualProgram(input: Record<string, string>, options: CompilerOptions = {}): ts.Program {
    const compilerHost: ts.CompilerHost = {
        fileExists: fileName => fileName in input || ts.sys.fileExists(fileName),
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: ts.getDefaultLibFileName,
        readFile: () => "",
        getNewLine: () => "\n",
        useCaseSensitiveFileNames: () => false,
        writeFile() {},

        getSourceFile(fileName) {
            if (fileName in input) {
                return ts.createSourceFile(fileName, input[fileName], ts.ScriptTarget.Latest, false);
            }

            let filePath: string | undefined;

            if (fileName.startsWith("lib.")) {
                const typeScriptDir = path.dirname(require.resolve("typescript"));
                filePath = path.join(typeScriptDir, fileName);
            }

            if (fileName.includes("language-extensions")) {
                const dtsName = fileName.replace(/(\.d)?(\.ts)$/, ".d.ts");
                filePath = path.resolve(dtsName);
            }

            if (filePath !== undefined) {
                if (libCache[fileName]) return libCache[fileName];
                const content = fs.readFileSync(filePath, "utf8");
                libCache[fileName] = ts.createSourceFile(filePath, content, ts.ScriptTarget.Latest, false);
                return libCache[fileName];
            }
        },
    };

    return ts.createProgram(Object.keys(input), options, compilerHost);
}

export interface TranspileVirtualProjectResult {
    diagnostics: ts.Diagnostic[];
    transpiledFiles: TranspiledFile[];
}

export function transpileVirtualProject(
    files: Record<string, string>,
    options: CompilerOptions = {}
): TranspileVirtualProjectResult {
    const program = createVirtualProgram(files, options);
    const collector = createEmitOutputCollector();
    const { diagnostics: transpileDiagnostics } = new Transpiler().emit({ program, writeFile: collector.writeFile });
    const diagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...transpileDiagnostics,
    ]);

    return { diagnostics: [...diagnostics], transpiledFiles: collector.files };
}

export interface TranspileStringResult {
    diagnostics: ts.Diagnostic[];
    file?: TranspiledFile;
}

export function transpileString(main: string, options: CompilerOptions = {}): TranspileStringResult {
    const { diagnostics, transpiledFiles } = transpileVirtualProject({ "main.ts": main }, options);
    return {
        diagnostics,
        file: transpiledFiles.find(({ sourceFiles }) => sourceFiles.some(f => f.fileName === "main.ts")),
    };
}
