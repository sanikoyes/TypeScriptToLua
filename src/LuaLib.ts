import * as path from "path";
import { EmitHost } from "./transpilation";

export enum LuaLibFeature {
    ArrayConcat = "ArrayConcat",
    ArrayEntries = "ArrayEntries",
    ArrayEvery = "ArrayEvery",
    ArrayFilter = "ArrayFilter",
    ArrayForEach = "ArrayForEach",
    ArrayFind = "ArrayFind",
    ArrayFindIndex = "ArrayFindIndex",
    ArrayIncludes = "ArrayIncludes",
    ArrayIndexOf = "ArrayIndexOf",
    ArrayIsArray = "ArrayIsArray",
    ArrayJoin = "ArrayJoin",
    ArrayMap = "ArrayMap",
    ArrayPush = "ArrayPush",
    ArrayReduce = "ArrayReduce",
    ArrayReduceRight = "ArrayReduceRight",
    ArrayReverse = "ArrayReverse",
    ArrayShift = "ArrayShift",
    ArrayUnshift = "ArrayUnshift",
    ArraySort = "ArraySort",
    ArraySlice = "ArraySlice",
    ArraySome = "ArraySome",
    ArraySplice = "ArraySplice",
    ArrayToObject = "ArrayToObject",
    ArrayFlat = "ArrayFlat",
    ArrayFlatMap = "ArrayFlatMap",
    ArraySetLength = "ArraySetLength",
    Class = "Class",
    ClassExtends = "ClassExtends",
    CloneDescriptor = "CloneDescriptor",
    Decorate = "Decorate",
    DecorateParam = "DecorateParam",
    Delete = "Delete",
    DelegatedYield = "DelegatedYield",
    Error = "Error",
    FunctionBind = "FunctionBind",
    Generator = "Generator",
    InstanceOf = "InstanceOf",
    InstanceOfObject = "InstanceOfObject",
    Iterator = "Iterator",
    Map = "Map",
    MathAtan2 = "MathAtan2",
    New = "New",
    Number = "Number",
    NumberIsFinite = "NumberIsFinite",
    NumberIsNaN = "NumberIsNaN",
    NumberToString = "NumberToString",
    ObjectAssign = "ObjectAssign",
    ObjectDefineProperty = "ObjectDefineProperty",
    ObjectEntries = "ObjectEntries",
    ObjectFromEntries = "ObjectFromEntries",
    ObjectGetOwnPropertyDescriptor = "ObjectGetOwnPropertyDescriptor",
    ObjectGetOwnPropertyDescriptors = "ObjectGetOwnPropertyDescriptors",
    ObjectKeys = "ObjectKeys",
    ObjectRest = "ObjectRest",
    ObjectValues = "ObjectValues",
    ParseFloat = "ParseFloat",
    ParseInt = "ParseInt",
    Set = "Set",
    SetDescriptor = "SetDescriptor",
    WeakMap = "WeakMap",
    WeakSet = "WeakSet",
    SourceMapTraceBack = "SourceMapTraceBack",
    Spread = "Spread",
    StringAccess = "StringAccess",
    StringCharAt = "StringCharAt",
    StringCharCodeAt = "StringCharCodeAt",
    StringConcat = "StringConcat",
    StringEndsWith = "StringEndsWith",
    StringIncludes = "StringIncludes",
    StringPadEnd = "StringPadEnd",
    StringPadStart = "StringPadStart",
    StringReplace = "StringReplace",
    StringSlice = "StringSlice",
    StringSplit = "StringSplit",
    StringStartsWith = "StringStartsWith",
    StringSubstr = "StringSubstr",
    StringSubstring = "StringSubstring",
    StringTrim = "StringTrim",
    StringTrimEnd = "StringTrimEnd",
    StringTrimStart = "StringTrimStart",
    Symbol = "Symbol",
    SymbolRegistry = "SymbolRegistry",
    TypeOf = "TypeOf",
    Unpack = "Unpack",
}

/* eslint-disable @typescript-eslint/naming-convention */
const luaLibDependencies: Partial<Record<LuaLibFeature, LuaLibFeature[]>> = {
    ArrayConcat: [LuaLibFeature.ArrayIsArray],
    ArrayFlat: [LuaLibFeature.ArrayConcat, LuaLibFeature.ArrayIsArray],
    ArrayFlatMap: [LuaLibFeature.ArrayConcat, LuaLibFeature.ArrayIsArray],
    Decorate: [LuaLibFeature.CloneDescriptor],
    Delete: [LuaLibFeature.ObjectGetOwnPropertyDescriptors],
    Error: [LuaLibFeature.New, LuaLibFeature.Class],
    FunctionBind: [LuaLibFeature.Unpack],
    Generator: [LuaLibFeature.Symbol],
    InstanceOf: [LuaLibFeature.Symbol],
    Iterator: [LuaLibFeature.Symbol],
    ObjectDefineProperty: [LuaLibFeature.CloneDescriptor, LuaLibFeature.SetDescriptor],
    ObjectFromEntries: [LuaLibFeature.Iterator, LuaLibFeature.Symbol],
    Map: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol, LuaLibFeature.Class],
    Set: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol, LuaLibFeature.Class],
    WeakMap: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol, LuaLibFeature.Class],
    WeakSet: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol, LuaLibFeature.Class],
    Spread: [LuaLibFeature.Iterator, LuaLibFeature.Unpack],
    StringSplit: [LuaLibFeature.StringSubstring, LuaLibFeature.StringAccess],
    SymbolRegistry: [LuaLibFeature.Symbol],
};
/* eslint-enable @typescript-eslint/naming-convention */

export function loadLuaLibFeatures(features: Iterable<LuaLibFeature>, emitHost: EmitHost): string {
    let result = "";

    const loadedFeatures = new Set<LuaLibFeature>();

    function load(feature: LuaLibFeature): void {
        if (loadedFeatures.has(feature)) return;
        loadedFeatures.add(feature);

        const dependencies = luaLibDependencies[feature];
        if (dependencies) {
            dependencies.forEach(load);
        }

        const featurePath = path.resolve(__dirname, `../dist/lualib/${feature}.lua`);
        const luaLibFeature = emitHost.readFile(featurePath);
        if (luaLibFeature !== undefined) {
            result += luaLibFeature + "\n";
        } else {
            throw new Error(`Could not load lualib feature from '${featurePath}'`);
        }
    }

    for (const feature of features) {
        load(feature);
    }

    return `local _ENV = setmetatable({}, { __index = _G })
pcall(function() setfenv(1, _ENV) end)

${result}

return _ENV
`;
}

let luaLibBundleContent: string;
export function getLuaLibBundle(emitHost: EmitHost): string {
    if (luaLibBundleContent === undefined) {
        const lualibPath = path.resolve(__dirname, "../dist/lualib/lualib_bundle.lua");
        const result = emitHost.readFile(lualibPath);
        if (result !== undefined) {
            luaLibBundleContent = result;
        } else {
            throw new Error(`Could not load lualib bundle from '${lualibPath}'`);
        }
    }

    return luaLibBundleContent;
}
