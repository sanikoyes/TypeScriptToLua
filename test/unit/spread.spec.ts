import * as tstl from "../../src";
import * as util from "../util";
import { formatCode } from "../util";

// TODO: Make some utils for testing other targets
const expectUnpack: util.TapCallback = builder => expect(builder.getMainLuaCodeChunk()).toMatch(/[^.]unpack\(/);
const expectTableUnpack: util.TapCallback = builder => expect(builder.getMainLuaCodeChunk()).toContain("table.unpack");
const expectLualibUnpack: util.TapCallback = builder => expect(builder.getMainLuaCodeChunk()).toContain("__TS__Unpack");

const arrayLiteralCases = [
    "1, 2, ...[3, 4, 5]",
    "...[1, 2], 3, 4, 5",
    "1, ...[[2]], 3",
    "...[1, 2, 3], 4, ...[5, 6]",
    "1, 2, ...[3, 4], ...[5, 6]",
];

describe.each(["function call", "array literal"] as const)("in %s", kind => {
    const factory = (code: string) => (kind === "function call" ? `((...args: any[]) => args)(${code})` : `[${code}]`);

    test.each(arrayLiteralCases)("of array literal (%p)", expression => {
        util.testExpression(factory(expression)).expectToMatchJsResult();
    });

    test.each(arrayLiteralCases)("of multi return call (%p)", expression => {
        util.testFunction`
            function tuple(...args: any[]) {
                return $multi(...args);
            }

            return ${factory(`...tuple(${expression})`)};
        `
            .withLanguageExtensions()
            .expectToMatchJsResult();
    });

    test("of multiple string literals", () => {
        util.testExpression(factory('..."spread", ..."string"')).expectToMatchJsResult();
    });

    test.each(["", "string", "string with spaces", "string 1 2 3"])("of string literal (%p)", str => {
        util.testExpression(factory(`...${formatCode(str)}`)).expectToMatchJsResult();
    });

    test("of iterable", () => {
        util.testFunction`
            const it = {
                i: -1,
                [Symbol.iterator]() {
                    return this;
                },
                next() {
                    ++this.i;
                    return {
                        value: 2 ** this.i,
                        done: this.i == 9,
                    }
                }
            };

            return ${factory("...it")};
        `.expectToMatchJsResult();
    });
});

describe("in function call", () => {
    util.testEachVersion(
        undefined,
        () => util.testFunction`
            function foo(a: number, b: number, ...rest: number[]) {
                return { a, b, rest }
            }
            const array = [0, 1, 2, 3] as const;
            return foo(...array);
        `,
        {
            [tstl.LuaTarget.Universal]: builder => builder.tap(expectLualibUnpack).expectToMatchJsResult(),
            [tstl.LuaTarget.LuaJIT]: builder => builder.tap(expectUnpack),
            [tstl.LuaTarget.Lua51]: builder => builder.tap(expectUnpack).expectToMatchJsResult(),
            [tstl.LuaTarget.Lua52]: builder => builder.tap(expectTableUnpack).expectToMatchJsResult(),
            [tstl.LuaTarget.Lua53]: builder => builder.tap(expectTableUnpack).expectToMatchJsResult(),
            [tstl.LuaTarget.Lua54]: builder => builder.tap(expectTableUnpack).expectToMatchJsResult(),
        }
    );
});

describe("in array literal", () => {
    util.testEachVersion(undefined, () => util.testExpression`[...[0, 1, 2]]`, {
        [tstl.LuaTarget.Universal]: builder => builder.tap(expectLualibUnpack).expectToMatchJsResult(),
        [tstl.LuaTarget.LuaJIT]: builder => builder.tap(expectUnpack),
        [tstl.LuaTarget.Lua51]: builder => builder.tap(expectUnpack).expectToMatchJsResult(),
        [tstl.LuaTarget.Lua52]: builder => builder.tap(expectTableUnpack).expectToMatchJsResult(),
        [tstl.LuaTarget.Lua53]: builder => builder.tap(expectTableUnpack).expectToMatchJsResult(),
        [tstl.LuaTarget.Lua54]: builder => builder.tap(expectTableUnpack).expectToMatchJsResult(),
    });

    test("of array literal /w OmittedExpression", () => {
        util.testFunction`
            const array = [1, 2, ...[3], 5, , 6];
            return { a: array[0], b: array[1], c: array[2], d: array[3] };
        `.expectToMatchJsResult();
    });
});

describe("in object literal", () => {
    test.each([
        "{ x: false, ...{ x: true, y: true } }",
        "{ ...{ x: true, y: true } }",
        "{ ...{ x: true }, ...{ y: true, z: true } }",
        "{ ...{ x: false }, x: true }",
        "{ ...{ x: false }, x: false, ...{ x: true } }",
    ])("of object literal (%p)", expression => {
        util.testExpression(expression).expectToMatchJsResult();
    });

    test("of object reference", () => {
        util.testFunction`
            const object = { x: 0, y: 1 };
            const result = { ...object, z: 2 };
            return { object, result };
        `.expectToMatchJsResult();
    });

    test.each([
        ["literal", "const object = { ...[0, 1, 2] };"],
        ["reference", "const array = [0, 1, 2]; const object = { ...array };"],
    ])("of array %p", (_name, expressionToCreateObject) => {
        util.testFunction`
            ${expressionToCreateObject}
            return { "0": object[0], "1": object[1], "2": object[2] };
        `.expectToMatchJsResult();
    });
});

describe("vararg spread optimization", () => {
    test("basic use", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                return pick(...args);
            }
            return test("a", "b", "c");
        `
            .expectLuaToMatchSnapshot()
            .expectToMatchJsResult();
    });

    test("body-less arrow function", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            const test = (...args: string[]) => pick(...args);
            return test("a", "b", "c");
        `
            .expectLuaToMatchSnapshot()
            .expectToMatchJsResult();
    });

    test("if statement", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                if (true) {
                    return pick(...args);
                }
            }
            return test("a", "b", "c");
        `
            .expectLuaToMatchSnapshot()
            .expectToMatchJsResult();
    });

    test("loop statement", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                do {
                    return pick(...args);
                } while (false);
            }
            return test("a", "b", "c");
        `
            .expectLuaToMatchSnapshot()
            .expectToMatchJsResult();
    });

    test("block statement", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                let result: string;
                {
                    result = pick(...args);
                }
                return result;
            }
            return test("a", "b", "c");
        `
            .expectLuaToMatchSnapshot()
            .expectToMatchJsResult();
    });

    test("finally clause", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                try {
                    throw "foobar";
                } catch {
                } finally {
                    return pick(...args);
                }
            }
            return test("a" ,"b", "c");
        `
            .expectLuaToMatchSnapshot()
            .expectToMatchJsResult();
    });

    test("$multi", () => {
        util.testFunction`
            function multi(...args: string[]) {
                return $multi(...args);
            }
            function test(...args: string[]) {
                return multi(...args)[1];
            }
            return test("a" ,"b", "c");
        `
            .withLanguageExtensions()
            .expectLuaToMatchSnapshot()
            .expectToEqual("b");
    });

    test("curry", () => {
        util.testFunction`
            function test<A extends any[]>(fn: (...args: A) => void, ...args: A) {
                return fn(...args);
            }
            return test((arg: string) => arg, "foobar");
        `
            .expectLuaToMatchSnapshot()
            .expectToMatchJsResult();
    });

    test("curry with indirect type", () => {
        util.testFunction`
            function test<A extends any[]>(obj: {fn: (...args: A) => void}, ...args: A) {
                const fn = obj.fn;
                return fn(...args);
            }
            return test({fn: (arg: string) => arg}, "foobar");
        `
            .expectLuaToMatchSnapshot()
            .expectToMatchJsResult();
    });

    test("function type declared inside scope", () => {
        util.testFunction`
            function test<A extends any[]>(...args: A) {
                const fn: (...args: A) => A[0] = (...args) => args[0];
                return fn(...args);
            }
            test("foobar");
        `
            .expectLuaToMatchSnapshot()
            .expectToMatchJsResult();
    });
});

describe("vararg spread de-optimization", () => {
    test("array modification", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                args[1] = "foobar";
                return pick(...args);
            }
            return test("a", "b", "c");
        `.expectToMatchJsResult();
    });

    test("array modification in hoisted function", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                hoisted();
                const result = pick(...args);
                function hoisted() { args[1] = "foobar"; }
                return result;
            }
            return test("a", "b", "c");
        `.expectToMatchJsResult();
    });

    test("array modification in secondary hoisted function", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                function triggersHoisted() { hoisted(); }
                triggersHoisted();
                const result = pick(...args);
                function hoisted() { args[1] = "foobar"; }
                return result;
            }
            return test("a", "b", "c");
        `.expectToMatchJsResult();
    });
});

describe("vararg spread in IIFE", () => {
    test("comma operator", () => {
        util.testFunction`
            function dummy() { return "foobar"; }
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                return (dummy(), pick(...args));
            }
            return test("a", "b", "c");
        `.expectToMatchJsResult();
    });

    test("assignment expression", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                let x: string;
                return (x = pick(...args));
            }
            return test("a", "b", "c");
        `.expectToMatchJsResult();
    });

    test("destructured assignment expression", () => {
        util.testFunction`
            function pick(...args: string[]) { return [args[1]]; }
            function test(...args: string[]) {
                let x: string;
                return ([x] = pick(...args));
            }
            return test("a", "b", "c");
        `.expectToMatchJsResult();
    });

    test("property-access assignment expression", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                let x: {val?: string} = {};
                return (x.val = pick(...args));
            }
            return test("a", "b", "c");
        `.expectToMatchJsResult();
    });

    test("binary compound assignment", () => {
        util.testFunction`
            function pick(...args: number[]) { return args[1]; }
            function test(...args: number[]) {
                let x = 1;
                return x += pick(...args);
            }
            return test(1, 2, 3);
        `.expectToMatchJsResult();
    });

    test("postfix unary compound assignment", () => {
        util.testFunction`
            function pick(...args: number[]) { return args[1]; }
            function test(...args: number[]) {
                let x = [7, 8, 9];
                return x[pick(...args)]++;
            }
            return test(1, 2, 3);
        `.expectToMatchJsResult();
    });

    test("prefix unary compound assignment", () => {
        util.testFunction`
            function pick(...args: number[]) { return args[1]; }
            function test(...args: number[]) {
                let x = [7, 8, 9];
                return ++x[pick(...args)];
            }
            return test(1, 2, 3);
        `.expectToMatchJsResult();
    });

    test("try clause", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                try {
                    return pick(...args)
                } catch {}
            }
            return test("a" ,"b", "c");
        `.expectToMatchJsResult();
    });

    test("catch clause", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                try {
                    throw "foobar";
                } catch {
                    return pick(...args)
                }
            }
            return test("a" ,"b", "c");
        `.expectToMatchJsResult();
    });

    test("class expression", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            function test(...args: string[]) {
                const fooClass = class Foo { foo = pick(...args); };
                return new fooClass().foo;
            }
            return test("a" ,"b", "c");
        `.expectToMatchJsResult();
    });

    test("self-referencing function expression", () => {
        util.testFunction`
            function pick(...args: string[]) { return args[1]; }
            const test = function testName(...args: string[]) {
                return \`\${typeof testName}:\${pick(...args)}\`;
            }
            return test("a" ,"b", "c");
        `.expectToMatchJsResult();
    });

    test("method indirect access (access args)", () => {
        util.testFunction`
            const obj = { $method: () => obj.arg, arg: "foobar" };
            function getObj(...args: string[]) { obj.arg = args[1]; return obj; }
            function test(...args: string[]) {
                return getObj(...args).$method();
            }
            return test("a" ,"b", "c");
        `.expectToMatchJsResult();
    });

    test("method indirect access (method args)", () => {
        util.testFunction`
            const obj = { $pick: (...args: string[]) => args[1] };
            function getObj() { return obj; }
            function test(...args: string[]) {
                return getObj().$pick(...args);
            }
            return test("a" ,"b", "c");
        `.expectToMatchJsResult();
    });

    test("tagged template method indirect access", () => {
        util.testFunction`
            const obj = { $tag: (t: TemplateStringsArray, ...args: string[]) => args[1] };
            function getObj() { return obj; }
            function pick(...args: string[]): string { return args[1]; }
            function test(...args: string[]) {
                return getObj().$tag\`FOO\${pick(...args)}BAR\`;
            }
            return test("a" ,"b", "c");
        `.expectToMatchJsResult();
    });
});
