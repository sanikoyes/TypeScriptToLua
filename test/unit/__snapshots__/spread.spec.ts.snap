// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vararg spread optimization $multi 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function multi(self, ...)
        return ...
    end
    local function test(self, ...)
        return select(
            2,
            multi(nil, ...)
        )
    end
    return test(nil, \\"a\\", \\"b\\", \\"c\\")
end
return ____exports"
`;

exports[`vararg spread optimization basic use 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function pick(self, ...)
        local args = {...}
        return args[2]
    end
    local function test(self, ...)
        return pick(nil, ...)
    end
    return test(nil, \\"a\\", \\"b\\", \\"c\\")
end
return ____exports"
`;

exports[`vararg spread optimization block statement 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function pick(self, ...)
        local args = {...}
        return args[2]
    end
    local function test(self, ...)
        local result
        do
            result = pick(nil, ...)
        end
        return result
    end
    return test(nil, \\"a\\", \\"b\\", \\"c\\")
end
return ____exports"
`;

exports[`vararg spread optimization body-less arrow function 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function pick(self, ...)
        local args = {...}
        return args[2]
    end
    local function test(____, ...)
        return pick(nil, ...)
    end
    return test(nil, \\"a\\", \\"b\\", \\"c\\")
end
return ____exports"
`;

exports[`vararg spread optimization curry 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function test(self, fn, ...)
        return fn(nil, ...)
    end
    return test(
        nil,
        function(____, arg) return arg end,
        \\"foobar\\"
    )
end
return ____exports"
`;

exports[`vararg spread optimization curry with indirect type 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function test(self, obj, ...)
        local fn = obj.fn
        return fn(nil, ...)
    end
    return test(
        nil,
        {
            fn = function(____, arg) return arg end
        },
        \\"foobar\\"
    )
end
return ____exports"
`;

exports[`vararg spread optimization finally clause 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function pick(self, ...)
        local args = {...}
        return args[2]
    end
    local function test(self, ...)
        do
            pcall(
                function()
                    error(\\"foobar\\", 0)
                end
            )
            do
                return pick(nil, ...)
            end
        end
    end
    return test(nil, \\"a\\", \\"b\\", \\"c\\")
end
return ____exports"
`;

exports[`vararg spread optimization function type declared inside scope 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function test(self, ...)
        local function fn(____, ...)
            local args = {...}
            return args[1]
        end
        return fn(nil, ...)
    end
    test(nil, \\"foobar\\")
end
return ____exports"
`;

exports[`vararg spread optimization if statement 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function pick(self, ...)
        local args = {...}
        return args[2]
    end
    local function test(self, ...)
        if true then
            return pick(nil, ...)
        end
    end
    return test(nil, \\"a\\", \\"b\\", \\"c\\")
end
return ____exports"
`;

exports[`vararg spread optimization loop statement 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function pick(self, ...)
        local args = {...}
        return args[2]
    end
    local function test(self, ...)
        repeat
            do
                return pick(nil, ...)
            end
        until not false
    end
    return test(nil, \\"a\\", \\"b\\", \\"c\\")
end
return ____exports"
`;
